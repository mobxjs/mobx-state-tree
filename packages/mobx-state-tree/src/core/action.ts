import { action as mobxAction } from "mobx"
import {
    getStateTreeNode,
    fail,
    argsToArray,
    IDisposer,
    getRoot,
    EMPTY_ARRAY,
    ObjectNode,
    HookNames,
    IAnyStateTreeNode
} from "../internal"

export type IMiddlewareEventType =
    | "action"
    | "flow_spawn"
    | "flow_resume"
    | "flow_resume_error"
    | "flow_return"
    | "flow_throw"
// | "task_spawn TODO, see #273"

export type IMiddlewareEvent = {
    type: IMiddlewareEventType
    name: string
    id: number
    parentId: number
    rootId: number
    context: IAnyStateTreeNode
    tree: IAnyStateTreeNode
    args: any[]
}

export type IMiddleware = {
    handler: IMiddlewareHandler
    includeHooks: boolean
}
export type IMiddlewareHandler = (
    actionCall: IMiddlewareEvent,
    next: (actionCall: IMiddlewareEvent, callback?: (value: any) => any) => void,
    abort: (value: any) => void
) => any

let nextActionId = 1
let currentActionContext: IMiddlewareEvent | null = null

export function getNextActionId() {
    return nextActionId++
}

// TODO: optimize away entire action context if there is no middleware in tree?
export function runWithActionContext(context: IMiddlewareEvent, fn: Function) {
    const node = getStateTreeNode(context.context)
    const baseIsRunningAction = node._isRunningAction
    const prevContext = currentActionContext

    if (context.type === "action") {
        node.assertAlive()
    }

    node._isRunningAction = true
    currentActionContext = context
    try {
        return runMiddleWares(node, context, fn)
    } finally {
        currentActionContext = prevContext
        node._isRunningAction = baseIsRunningAction
    }
}

export function getActionContext(): IMiddlewareEvent {
    if (!currentActionContext) return fail("Not running an action!")
    return currentActionContext
}

export function createActionInvoker<T extends Function>(
    target: IAnyStateTreeNode,
    name: string,
    fn: T
) {
    const res = function() {
        const id = getNextActionId()
        return runWithActionContext(
            {
                type: "action",
                name,
                id,
                args: argsToArray(arguments),
                context: target,
                tree: getRoot(target),
                rootId: currentActionContext ? currentActionContext.rootId : id,
                parentId: currentActionContext ? currentActionContext.id : 0
            },
            fn
        )
    }
    ;(res as any)._isMSTAction = true
    return res
}

/**
 * Middleware can be used to intercept any action is invoked on the subtree where it is attached.
 * If a tree is protected (by default), this means that any mutation of the tree will pass through your middleware.
 *
 * For more details, see the [middleware docs](docs/middleware.md)
 *
 * @export
 * @param {IStateTreeNode} target
 * @param {(action: IRawActionCall, next: (call: IRawActionCall) => any) => any} middleware
 * @returns {IDisposer}
 */
export function addMiddleware(
    target: IAnyStateTreeNode,
    handler: IMiddlewareHandler,
    includeHooks: boolean = true
): IDisposer {
    const node = getStateTreeNode(target)
    if (process.env.NODE_ENV !== "production") {
        if (!node.isProtectionEnabled)
            console.warn(
                "It is recommended to protect the state tree before attaching action middleware, as otherwise it cannot be guaranteed that all changes are passed through middleware. See `protect`"
            )
    }
    return node.addMiddleWare(handler, includeHooks)
}

export function decorate<T extends Function>(middleware: IMiddlewareHandler, fn: T): T
/**
 * Binds middleware to a specific action
 *
 * @example
 * type.actions(self => {
 *   function takeA____() {
 *       self.toilet.donate()
 *       self.wipe()
 *       self.wipe()
 *       self.toilet.flush()
 *   }
 *   return {
 *     takeA____: decorate(atomic, takeA____)
 *   }
 * })
 *
 * @export
 * @template T
 * @param {IMiddlewareHandler} handler
 * @param Function} fn
 * @returns the original function
 */
export function decorate<T extends Function>(handler: IMiddlewareHandler, fn: any) {
    const middleware: IMiddleware = { handler, includeHooks: true }
    if (fn.$mst_middleware) fn.$mst_middleware.push(middleware)
    else fn.$mst_middleware = [middleware]
    return fn
}

function collectMiddlewares(
    node: ObjectNode,
    baseCall: IMiddlewareEvent,
    fn: Function
): IMiddleware[] {
    let middlewares: IMiddleware[] = (fn as any).$mst_middleware || EMPTY_ARRAY
    let n: ObjectNode | null = node
    // Find all middlewares. Optimization: cache this?
    while (n) {
        if (n.middlewares) middlewares = middlewares.concat(n.middlewares)
        n = n.parent
    }
    return middlewares
}

function runMiddleWares(node: ObjectNode, baseCall: IMiddlewareEvent, originalFn: Function): any {
    const middlewares = collectMiddlewares(node, baseCall, originalFn)
    // Short circuit
    if (!middlewares.length) return mobxAction(originalFn).apply(null, baseCall.args)
    let index = 0
    let result: any = null

    function runNextMiddleware(call: IMiddlewareEvent): any {
        const middleware = middlewares[index++]
        const handler = middleware && middleware.handler
        let nextInvoked = false
        let abortInvoked = false

        function next(call: IMiddlewareEvent): void
        function next(call: IMiddlewareEvent, callback: (value: any) => any): void
        function next(call: IMiddlewareEvent, callback?: (value: any) => any) {
            nextInvoked = true
            // the result can contain
            // - the non manipulated return value from an action
            // - the non manipulated abort value
            // - one of the above but manipulated through the callback function
            if (callback) {
                result = callback(runNextMiddleware(call) || result)
            } else {
                result = runNextMiddleware(call)
            }
        }
        function abort(value: any) {
            abortInvoked = true
            // overwrite the result
            // can be manipulated through middlewares earlier in the queue using the callback fn
            result = value
        }
        const invokeHandler = () => {
            handler(call, next, abort)
            if (process.env.NODE_ENV !== "production") {
                if (!nextInvoked && !abortInvoked) {
                    const node = getStateTreeNode(call.tree)
                    fail(
                        `Neither the next() nor the abort() callback within the middleware ${
                            handler.name
                        } for the action: "${call.name}" on the node: ${
                            node.type.name
                        } was invoked.`
                    )
                }
                if (nextInvoked && abortInvoked) {
                    const node = getStateTreeNode(call.tree)
                    fail(
                        `The next() and abort() callback within the middleware ${
                            handler.name
                        } for the action: "${call.name}" on the node: ${
                            node.type.name
                        } were invoked.`
                    )
                }
            }
            return result
        }

        if (handler && middleware.includeHooks) {
            return invokeHandler()
        } else if (handler && !middleware.includeHooks) {
            if ((HookNames as any)[call.name]) return runNextMiddleware(call)
            return invokeHandler()
        } else {
            return mobxAction(originalFn).apply(null, call.args)
        }
    }
    return runNextMiddleware(baseCall)
}
