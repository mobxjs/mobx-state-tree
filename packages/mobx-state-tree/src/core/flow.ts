/** @hidden */
declare const $flowYield: unique symbol

/** @hidden */
export interface FlowYield {
    // fake, only for typing
    [$flowYield]: undefined
}

/** @hidden */
declare const $flowReturn: unique symbol

/** @hidden */
export interface FlowReturn<T> {
    // fake, only for typing
    [$flowReturn]: T
}

// we skip promises that are the result of yielding promises (except if they use flowReturn)
/** @hidden */
export type FlowReturnType<R> = IfAllAreFlowYieldThenVoid<
    R extends FlowReturn<infer FR>
        ? FR extends Promise<infer FRP>
            ? FRP
            : FR
        : R extends Promise<any>
        ? FlowYield
        : R
>

// we extract yielded promises from the return type
/** @hidden */
export type IfAllAreFlowYieldThenVoid<R> = Exclude<R, FlowYield> extends never
    ? void
    : Exclude<R, FlowYield>

/**
 * See [asynchronous actions](https://github.com/mobxjs/mobx-state-tree/blob/master/docs/async-actions.md).
 *
 * @returns The flow as a promise.
 */
export function flow<R, Args extends any[]>(
    generator: (...args: Args) => IterableIterator<R>
): (...args: Args) => Promise<FlowReturnType<R>> {
    return createFlowSpawner(generator.name, generator) as any
}

/**
 *  Used for TypeScript to make flows that return a promise return the actual promise result.
 *
 * @param val
 * @returns
 */
export function castFlowReturn<T>(val: T): FlowReturn<T> {
    return val as any
}

/**
 * @internal
 * @hidden
 */
export function createFlowSpawner(name: string, generator: Function) {
    const spawner = function flowSpawner(this: any) {
        // Implementation based on https://github.com/tj/co/blob/master/index.js
        const runId = getNextActionId()
        const baseContext = getActionContext()
        const args = arguments

        function wrap(fn: any, type: IMiddlewareEventType, arg: any) {
            fn.$mst_middleware = (spawner as any).$mst_middleware // pick up any middleware attached to the flow
            runWithActionContext(
                {
                    name,
                    type,
                    id: runId,
                    args: [arg],
                    tree: baseContext.tree,
                    context: baseContext.context,
                    parentId: baseContext.id,
                    allParentIds: [...baseContext.allParentIds, baseContext.id],
                    rootId: baseContext.rootId
                },
                fn
            )
        }

        return new Promise(function(resolve, reject) {
            let gen: any
            const init = function asyncActionInit() {
                gen = generator.apply(null, arguments)
                onFulfilled(undefined) // kick off the flow
            }
            ;(init as any).$mst_middleware = (spawner as any).$mst_middleware

            runWithActionContext(
                {
                    name,
                    type: "flow_spawn",
                    id: runId,
                    args: argsToArray(args),
                    tree: baseContext.tree,
                    context: baseContext.context,
                    parentId: baseContext.id,
                    allParentIds: [...baseContext.allParentIds, baseContext.id],
                    rootId: baseContext.rootId
                },
                init
            )

            function onFulfilled(res: any) {
                let ret
                try {
                    // prettier-ignore
                    wrap((r: any) => { ret = gen.next(r) }, "flow_resume", res)
                } catch (e) {
                    // prettier-ignore
                    setImmediate(() => {
                        wrap((r: any) => { reject(e) }, "flow_throw", e)
                    })
                    return
                }
                next(ret)
                return
            }

            function onRejected(err: any) {
                let ret
                try {
                    // prettier-ignore
                    wrap((r: any) => { ret = gen.throw(r) }, "flow_resume_error", err) // or yieldError?
                } catch (e) {
                    // prettier-ignore
                    setImmediate(() => {
                        wrap((r: any) => { reject(e) }, "flow_throw", e)
                    })
                    return
                }
                next(ret)
            }

            function next(ret: any) {
                if (ret.done) {
                    // prettier-ignore
                    setImmediate(() => {
                        wrap((r: any) => { resolve(r) }, "flow_return", ret.value)
                    })
                    return
                }
                // TODO: support more type of values? See https://github.com/tj/co/blob/249bbdc72da24ae44076afd716349d2089b31c4c/index.js#L100
                if (!ret.value || typeof ret.value.then !== "function") {
                    // istanbul ignore next
                    throw fail("Only promises can be yielded to `async`, got: " + ret)
                }
                return ret.value.then(onFulfilled, onRejected)
            }
        })
    }
    return spawner
}

import {
    IMiddlewareEventType,
    runWithActionContext,
    getActionContext,
    getNextActionId
} from "./action"
import { fail, argsToArray } from "../utils"
