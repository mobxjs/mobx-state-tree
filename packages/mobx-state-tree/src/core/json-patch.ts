import { fail } from "../internal"

// https://tools.ietf.org/html/rfc6902
// http://jsonpatch.com/

export type IJsonPatch = {
    op: "replace" | "add" | "remove"
    path: string
    value?: any
}

export type IReversibleJsonPatch = IJsonPatch & {
    oldValue: any // This goes beyond JSON-patch, but makes sure each patch can be inverse applied
}

export function splitPatch(patch: IReversibleJsonPatch): [IJsonPatch, IJsonPatch] {
    if (!("oldValue" in patch)) fail(`Patches without \`oldValue\` field cannot be inversed`)
    return [stripPatch(patch), invertPatch(patch)]
}

export function stripPatch(patch: IReversibleJsonPatch): IJsonPatch {
    // strips `oldvalue` information from the patch, so that it becomes a patch conform the json-patch spec
    // this removes the ability to undo the patch
    switch (patch.op) {
        case "add":
            return { op: "add", path: patch.path, value: patch.value }
        case "remove":
            return { op: "remove", path: patch.path }
        case "replace":
            return { op: "replace", path: patch.path, value: patch.value }
    }
}

function invertPatch(patch: IReversibleJsonPatch): IJsonPatch {
    switch (patch.op) {
        case "add":
            return {
                op: "remove",
                path: patch.path
            }
        case "remove":
            return {
                op: "add",
                path: patch.path,
                value: patch.oldValue
            }
        case "replace":
            return {
                op: "replace",
                path: patch.path,
                value: patch.oldValue
            }
    }
}

// simple simple simple check
function isNumber(x: string): boolean {
    return typeof x === "number"
}

/**
 * escape slashes and backslashes
 * http://tools.ietf.org/html/rfc6901
 */
export function escapeJsonPath(str: string): string {
    if (isNumber(str) === true) {
        return "" + str
    }
    return str.replace(/~/g, "~1").replace(/\//g, "~0")
}

/**
 * unescape slashes and backslashes
 */
export function unescapeJsonPath(str: string): string {
    return str.replace(/~0/g, "/").replace(/~1/g, "~")
}

export function joinJsonPath(path: string[]): string {
    // `/` refers to property with an empty name, while `` refers to root itself!
    if (path.length === 0) return ""
    return "/" + path.map(escapeJsonPath).join("/")
}

export function splitJsonPath(path: string): string[] {
    // `/` refers to property with an empty name, while `` refers to root itself!
    const parts = path.split("/").map(unescapeJsonPath)

    // path '/a/b/c' -> a b c
    // path '../../b/c -> .. .. b c
    return parts[0] === "" ? parts.slice(1) : parts
}
