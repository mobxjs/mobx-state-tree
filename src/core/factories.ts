import {action} from "mobx"
import {extend, fail} from "../utils"
import {getNode, hasNode, NodeConstructor} from "./node"

export type IModel = {
    $treenode: any // Actually Node, but that should not be exposed to the public...
} & Object

export type IModelFactoryChecker = (snapshot: any) => boolean
export type IModelFactoryDispatcher = (snapshot?: any) => IModelFactory<any, any>
export interface IModelFactoryConstructor<S, T>{
    (snapshot?: S, env?: Object): T & IModel
}

export interface IModelFactory<S, T> extends IModelFactoryConstructor<S, T>{
    factoryName: string,
    kind: string
    is: IModelFactoryChecker
    irreducible: boolean
    dispatch: IModelFactoryDispatcher
    config: Object
    isModelFactory: boolean
}

export function createFactoryConstructor<S, T>(
    name: string,
    nodeClass: NodeConstructor,
    config,
    instanceCreator: () => any
): IModelFactoryConstructor<S, T>{
    let factory = extend(
        action(name, function(snapshot?: any, environment?: Object) {
            const instance = instanceCreator()
            const adm = new nodeClass(instance, environment, factory, config)
            if (arguments.length > 0)
                adm.applySnapshot(snapshot)
            return instance
        }),
        {
            config
        }
    )
    return factory
}

export function createFactory<S, T>(
    name: string,
    kind: string,
    is: IModelFactoryChecker,
    dispatch: IModelFactoryDispatcher,
    constructor: IModelFactoryConstructor<S, T>
): IModelFactory<S, T> {
    let factory = extend(
        constructor,
        {
            isModelFactory: true,
            factoryName: name,
            kind,
            is: (nodeOrSnapshot) => {
                let snapshot = isModel(nodeOrSnapshot) ? getNode(nodeOrSnapshot).snapshot : nodeOrSnapshot
                return is(snapshot)
            },
            dispatch: (nodeOrSnapshot) => {
                let snapshot = isModel(nodeOrSnapshot) ? getNode(nodeOrSnapshot).snapshot : nodeOrSnapshot
                return dispatch(snapshot)
            }
        }
    )
    return factory
}

export function isModelFactory(value: any): value is IModelFactory<any, any> {
    return typeof value === "function" && value.isModelFactory === true
}

export function getModelFactory(object: IModel): IModelFactory<any, any> {
    return getNode(object).factory
}

export function getChildModelFactory(object: IModel, child: string): IModelFactory<any, any> {
    return getNode(object).getChildFactory(child)
}

export function isModel(model: any): model is IModel {
    return hasNode(model)
}
