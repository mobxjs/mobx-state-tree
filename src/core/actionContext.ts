import { IAnyStateTreeNode, IMiddlewareEvent } from "../internal"
import { getCurrentActionContext } from "./action"

export interface IActionContext {
    /** Event name (action name for actions) */
    readonly name: string

    /** Event unique id */
    readonly id: number

    /** Parent action event object */
    readonly parentActionEvent: IMiddlewareEvent | undefined

    /** Event context (node where the action was invoked) */
    readonly context: IAnyStateTreeNode
    /** Event tree (root node of the node where the action was invoked) */
    readonly tree: IAnyStateTreeNode

    /** Event arguments in an array (action arguments for actions) */
    readonly args: any[]
}

/**
 * Returns the currently executing MST action context, or undefined if none.
 */
export function getRunningActionContext(): IActionContext | undefined {
    let current = getCurrentActionContext()
    while (current && current.type !== "action") {
        current = current.parentActionEvent
    }
    return current
}

function _isActionContextThisOrChildOf(
    actionContext: IActionContext,
    sameOrParent: number | IActionContext | IMiddlewareEvent,
    includeSame: boolean
) {
    const parentId = typeof sameOrParent === "number" ? sameOrParent : sameOrParent.id

    let current: IActionContext | IMiddlewareEvent | undefined = includeSame
        ? actionContext
        : actionContext.parentActionEvent
    while (current) {
        if (current.id === parentId) {
            return true
        }
        current = current.parentActionEvent
    }
    return false
}

/**
 * Returns if the given action context is a parent of this action context.
 */
export function isActionContextChildOf(
    actionContext: IActionContext,
    parent: number | IActionContext | IMiddlewareEvent
) {
    return _isActionContextThisOrChildOf(actionContext, parent, false)
}

/**
 * Returns if the given action context is this or a parent of this action context.
 */
export function isActionContextThisOrChildOf(
    actionContext: IActionContext,
    parentOrThis: number | IActionContext | IMiddlewareEvent
) {
    return _isActionContextThisOrChildOf(actionContext, parentOrThis, true)
}
