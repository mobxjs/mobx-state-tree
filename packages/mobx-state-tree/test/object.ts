import { unprotect } from "../src/core/mst-operations"
import {
    destroy,
    detach,
    onSnapshot,
    onPatch,
    onAction,
    applyPatch,
    applyAction,
    applySnapshot,
    getSnapshot,
    types
} from "../src"
import { autorun, reaction, observable } from "mobx"

const createTestFactories = () => {
    const Factory = types
        .model({
            to: "world"
        })
        .actions(self => {
            function setTo(to) {
                self.to = to
            }
            return {
                setTo
            }
        })
    const ComputedFactory = types
        .model({
            width: 100,
            height: 200
        })
        .views(self => ({
            get area() {
                return self.width * self.height
            }
        }))
    const ComputedFactory2 = types
        .model({
            props: types.map(types.number)
        })
        .views(self => ({
            get area() {
                return self.props.get("width")! * self.props.get("height")!
            }
        }))
        .actions(self => {
            function setWidth(value) {
                self.props.set("width", value)
            }
            function setHeight(value) {
                self.props.set("height", value)
            }
            return {
                setWidth,
                setHeight
            }
        })
    const BoxFactory = types.model({
        width: 0,
        height: 0
    })
    const ColorFactory = types.model({
        color: "#FFFFFF"
    })
    return { Factory, ComputedFactory, ComputedFactory2, BoxFactory, ColorFactory }
}
// === FACTORY TESTS ===
test("it should create a factory", () => {
    const { Factory } = createTestFactories()
    const instance = Factory.create()
    const snapshot = getSnapshot(instance)
    expect(snapshot).toEqual({ to: "world" })
    expect(Factory.create().toJSON!()).toEqual({ to: "world" }) // toJSON is there as shortcut for getSnapshot(), primarily for debugging convenience
    expect(Factory.create().toString()).toEqual("AnonymousModel@<root>")
})
test("it should restore the state from the snapshot", () => {
    const { Factory } = createTestFactories()
    expect(getSnapshot(Factory.create({ to: "universe" }))).toEqual({ to: "universe" })
})
// === SNAPSHOT TESTS ===
test("it should emit snapshots", () => {
    const { Factory } = createTestFactories()
    const doc = Factory.create()
    unprotect(doc)
    let snapshots: any[] = []
    onSnapshot(doc, snapshot => snapshots.push(snapshot))
    doc.to = "universe"
    expect(snapshots).toEqual([{ to: "universe" }])
})
test("it should apply snapshots", () => {
    const { Factory } = createTestFactories()
    const doc = Factory.create()
    applySnapshot(doc, { to: "universe" })
    expect(getSnapshot(doc)).toEqual({ to: "universe" })
})
test("it should apply and accept null value for types.maybe(complexType)", () => {
    const Item = types.model({
        value: types.string
    })
    const Model = types.model({
        item: types.maybe(Item)
    })
    const myModel = Model.create()
    applySnapshot(myModel, { item: { value: "something" } })
    applySnapshot(myModel, { item: null })
    expect(getSnapshot(myModel)).toEqual({ item: null })
})
test("it should return a snapshot", () => {
    const { Factory } = createTestFactories()
    const doc = Factory.create()
    expect(getSnapshot(doc)).toEqual({ to: "world" })
})
// === PATCHES TESTS ===
test("it should emit patches", () => {
    const { Factory } = createTestFactories()
    const doc = Factory.create()
    unprotect(doc)
    let patches: any[] = []
    onPatch(doc, patch => patches.push(patch))
    doc.to = "universe"
    expect(patches).toEqual([{ op: "replace", path: "/to", value: "universe" }])
})
test("it should apply a patch", () => {
    const { Factory } = createTestFactories()
    const doc = Factory.create()
    applyPatch(doc, { op: "replace", path: "/to", value: "universe" })
    expect(getSnapshot(doc)).toEqual({ to: "universe" })
})
test("it should apply patches", () => {
    const { Factory } = createTestFactories()
    const doc = Factory.create()
    applyPatch(doc, [
        { op: "replace", path: "/to", value: "mars" },
        { op: "replace", path: "/to", value: "universe" }
    ])
    expect(getSnapshot(doc)).toEqual({ to: "universe" })
})
test("it should stop listening to patches patches", () => {
    const { Factory } = createTestFactories()
    const doc = Factory.create()
    unprotect(doc)
    let patches: any[] = []
    let disposer = onPatch(doc, patch => patches.push(patch))
    doc.to = "universe"
    disposer()
    doc.to = "mweststrate"
    expect(patches).toEqual([{ op: "replace", path: "/to", value: "universe" }])
})
// === ACTIONS TESTS ===
test("it should call actions correctly", () => {
    const { Factory } = createTestFactories()
    const doc = Factory.create()
    doc.setTo("universe")
    expect(getSnapshot(doc)).toEqual({ to: "universe" })
})
test("it should emit action calls", () => {
    const { Factory } = createTestFactories()
    const doc = Factory.create()
    let actions: any[] = []
    onAction(doc, action => actions.push(action))
    doc.setTo("universe")
    expect(actions).toEqual([{ name: "setTo", path: "", args: ["universe"] }])
})
test("it should apply action call", () => {
    const { Factory } = createTestFactories()
    const doc = Factory.create()
    applyAction(doc, { name: "setTo", path: "", args: ["universe"] })
    expect(getSnapshot(doc)).toEqual({ to: "universe" })
})
test("it should apply actions calls", () => {
    const { Factory } = createTestFactories()
    const doc = Factory.create()
    applyAction(doc, [
        { name: "setTo", path: "", args: ["mars"] },
        { name: "setTo", path: "", args: ["universe"] }
    ])
    expect(getSnapshot(doc)).toEqual({ to: "universe" })
})
// === COMPUTED VALUES ===
test("it should have computed properties", () => {
    const { ComputedFactory } = createTestFactories()
    const doc = ComputedFactory.create()
    unprotect(doc)
    doc.width = 3
    doc.height = 2
    expect(doc.area).toEqual(6)
})
test("it should throw if a replaced object is read or written to", () => {
    const Todo = types
        .model({
            title: "test"
        })
        .actions(self => {
            function fn() {}
            return {
                fn
            }
        })
    const Store = types.model({
        todo: Todo
    })
    const s = Store.create({
        todo: { title: "3" }
    })
    unprotect(s)
    const todo = s.todo
    s.todo = Todo.create({ title: "4" })
    expect(s.todo.title).toBe("4")
    const err =
        "[mobx-state-tree] This object has died and is no longer part of a state tree. It cannot be used anymore. The object (of type 'AnonymousModel') used to live at '/todo'. It is possible to access the last snapshot of this object using 'getSnapshot', or to create a fresh copy using 'clone'. If you want to remove an object from the tree without killing it, use 'detach' instead."
    expect(() => {
        todo.fn()
    }).toThrowError(
        "[mobx-state-tree] AnonymousModel@<root>[dead] cannot be used anymore as it has died; it has been removed from a state tree. If you want to remove an element from a tree and let it live on, use 'detach' or 'clone' the value"
    )
    expect(() => {
        todo.title
    }).toThrowError(err)
    expect(() => {
        todo.title = "5"
    }).toThrowError(err)
})
// === COMPOSE FACTORY ===
test("it should compose factories", () => {
    const { BoxFactory, ColorFactory } = createTestFactories()
    const ComposedFactory = types.compose(BoxFactory, ColorFactory)
    expect(getSnapshot(ComposedFactory.create())).toEqual({ width: 0, height: 0, color: "#FFFFFF" })
})
test("it should compose factories with computed properties", () => {
    const { ComputedFactory2, ColorFactory } = createTestFactories()
    const ComposedFactory = types.compose(ColorFactory, ComputedFactory2)
    const store = ComposedFactory.create({ props: { width: 100, height: 200 } })
    expect(getSnapshot(store)).toEqual({ props: { width: 100, height: 200 }, color: "#FFFFFF" })
    expect(store.area).toBe(20000)
    expect(typeof store.setWidth).toBe("function")
    expect(typeof store.setHeight).toBe("function")
})
test("it should compose multiple types with computed properties", () => {
    const { ComputedFactory2, ColorFactory } = createTestFactories()
    const ComposedFactory = types.compose(ColorFactory, ComputedFactory2)
    const store = ComposedFactory.create({ props: { width: 100, height: 200 } })
    expect(getSnapshot(store)).toEqual({ props: { width: 100, height: 200 }, color: "#FFFFFF" })
    expect(store.area).toBe(20000)
    expect(typeof store.setWidth).toBe("function")
    expect(typeof store.setHeight).toBe("function")
})
test("methods get overridden by compose", () => {
    const A = types
        .model({
            count: types.optional(types.number, 0)
        })
        .actions(self => {
            function increment() {
                self.count += 1
            }
            return {
                increment
            }
        })
    const B = A.actions(self => ({
        increment() {
            self.count += 10
        }
    }))
    const store = B.create()
    expect(getSnapshot(store)).toEqual({ count: 0 })
    expect(store.count).toBe(0)
    store.increment()
    expect(store.count).toBe(10)
})
test("compose should add new props", () => {
    const A = types.model({
        count: types.optional(types.number, 0)
    })
    const B = A.props({
        called: types.optional(types.number, 0)
    })
    const store = B.create()
    expect(getSnapshot(store)).toEqual({ count: 0, called: 0 })
    expect(store.count).toBe(0)
})
test("models should expose their actions to be used in a composable way", () => {
    const A = types
        .model({
            count: types.optional(types.number, 0)
        })
        .actions(self => {
            function increment() {
                self.count += 1
            }
            return {
                increment
            }
        })
    const B = A.props({
        called: types.optional(types.number, 0)
    }).actions(self => {
        const baseIncrement = self.increment
        return {
            increment() {
                baseIncrement()
                self.called += 1
            }
        }
    })
    const store = B.create()
    expect(getSnapshot(store)).toEqual({ count: 0, called: 0 })
    expect(store.count).toBe(0)
    store.increment()
    expect(store.count).toBe(1)
    expect(store.called).toBe(1)
})
test("compose should be overwrite", () => {
    const A = types
        .model({
            name: "",
            alias: ""
        })
        .views(self => ({
            get displayName() {
                return self.alias || self.name
            }
        }))
    const B = A.props({
        type: ""
    }).views(self => ({
        get displayName() {
            return self.alias || self.name + self.type
        }
    }))
    const storeA = A.create({ name: "nameA", alias: "aliasA" })
    const storeB = B.create({ name: "nameB", alias: "aliasB", type: "typeB" })
    const storeC = B.create({ name: "nameC", type: "typeC" })
    expect(storeA.displayName).toBe("aliasA")
    expect(storeB.displayName).toBe("aliasB")
    expect(storeC.displayName).toBe("nameCtypeC")
})
// === TYPE CHECKS ===
test("it should check the type correctly", () => {
    const { Factory } = createTestFactories()
    const doc = Factory.create()
    expect(Factory.is(doc)).toEqual(true)
    expect(Factory.is([])).toEqual(false)
    expect(Factory.is({})).toEqual(true)
    expect(Factory.is({ to: "mars" })).toEqual(true)
    expect(Factory.is({ wrongKey: true })).toEqual(true)
    expect(Factory.is({ to: 3 })).toEqual(false)
})
if (process.env.NODE_ENV !== "production") {
    test("it should require complex fields to be present", () => {
        expect(
            types
                .model({
                    todo: types.model({})
                })
                .is({})
        ).toBe(false)
        expect(() =>
            types
                .model({
                    todo: types.model({})
                })
                .create()
        ).toThrow()
        expect(
            types
                .model({
                    todo: types.array(types.string)
                })
                .is({})
        ).toBe(false) // TBD: or true?
        expect(() =>
            types
                .model({
                    todo: types.array(types.string)
                })
                .create()
        ).toThrow()
        expect(
            types
                .model({
                    todo: types.map(types.string)
                })
                .is({})
        ).toBe(false)
        expect(() =>
            types
                .model({
                    todo: types.map(types.string)
                })
                .create()
        ).toThrow()
    })
}
// === VIEW FUNCTIONS ===
test("view functions should be tracked", () => {
    const model = types
        .model({
            x: 3
        })
        .views(self => ({
            doubler() {
                return self.x * 2
            }
        }))
        .create()
    unprotect(model)
    const values: any[] = []
    const d = autorun(() => {
        values.push(model.doubler())
    })
    model.x = 7
    expect(values).toEqual([6, 14])
})
test("view functions should not be allowed to change state", () => {
    const model = types
        .model({
            x: 3
        })
        .views(self => ({
            doubler() {
                self.x *= 2
            }
        }))
        .actions(self => {
            function anotherDoubler() {
                self.x *= 2
            }
            return {
                anotherDoubler
            }
        })
        .create()
    expect(() => model.doubler()).toThrow()
    model.anotherDoubler()
    expect(model.x).toBe(6)
})
test("it should consider primitives as proposed defaults", () => {
    const now = new Date()
    const Todo = types.model({
        id: 0,
        name: "Hello world",
        done: false,
        createdAt: now
    })
    const doc = Todo.create()
    expect(getSnapshot(doc)).toEqual({
        id: 0,
        name: "Hello world",
        done: false,
        createdAt: now.getTime()
    })
})
test("it should throw if a non-primitive value is provided and no default can be created", () => {
    expect(() => {
        const Todo = types.model({
            complex: {
                a: 1,
                b: 2
            }
        })
    }).toThrow()
})
if (process.env.NODE_ENV !== "production") {
    test("it should not be possible to remove a node from a parent if it is required, see ", () => {
        const A = types.model("A", { x: 3 })
        const B = types.model("B", { a: A })
        const b = B.create({ a: { x: 7 } })
        unprotect(b)
        expect(() => {
            detach(b.a)
        }).toThrowError(/Error while converting `null` to `A`/)
        expect(() => {
            destroy(b.a)
        }).toThrowError(/Error while converting `null` to `A`/)
    })
}
test("it should be possible to share states between views and actions using enhance", () => {
    const A = types.model({}).extend(self => {
        const localState = observable.box(3)
        return {
            views: {
                get x() {
                    return localState.get()
                }
            },
            actions: {
                setX(value) {
                    localState.set(value)
                }
            }
        }
    })
    let x = 0
    let a = A.create()
    const d = reaction(
        () => a.x,
        v => {
            x = v
        }
    )
    a.setX(7)
    expect(a.x).toBe(7)
    expect(x).toBe(7)
    d()
})
test("It should throw if any other key is returned from extend", () => {
    const A = types.model({}).extend(() => ({ stuff() {} } as any)) // TODO: fix typing
    expect(() => A.create()).toThrowError(/stuff/)
})
