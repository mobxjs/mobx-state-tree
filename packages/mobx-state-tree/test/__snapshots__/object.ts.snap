// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should warn if a replaced object is read or written to 1`] = `
Array [
  Array [
    "[mobx-state-tree][error] You are trying to read or write to an object that is no longer part of a state tree. (Object type was 'Todo'). Either detach nodes first, or don't use objects after removing / replacing them in the tree. Use setLivelynessChecking(\\"error\\") to simplify debugging this error.",
  ],
  Array [
    "[mobx-state-tree][error] You are trying to read or write to an object that is no longer part of a state tree. (Object type was 'Todo'). Either detach nodes first, or don't use objects after removing / replacing them in the tree. Use setLivelynessChecking(\\"error\\") to simplify debugging this error.",
  ],
  Array [
    "[mobx-state-tree][error] You are trying to read or write to an object that is no longer part of a state tree. (Object type was 'Todo'). Either detach nodes first, or don't use objects after removing / replacing them in the tree. Use setLivelynessChecking(\\"error\\") to simplify debugging this error.",
  ],
  Array [
    "[mobx-state-tree][error] You are trying to read or write to an object that is no longer part of a state tree. (Object type was 'Todo'). Either detach nodes first, or don't use objects after removing / replacing them in the tree. Use setLivelynessChecking(\\"error\\") to simplify debugging this error.",
  ],
]
`;
